generate/src/contract/events.rs

```diff
@@ -247,13 +247,141 @@ fn expand_all_events(cx: &Context) -> TokenStream {
+                quote! {
+                    #struct_name(self::events::#struct_name)
+                }
+            })
+            .collect::<Vec<_>>()
+    };
+
+    quote! {
+        /// A contract event.
+        pub enum Event {
+            #( #variants, )*
+        }
+    }
+}
+
+/// Expands the `ParseLog` implementation for the event enum.
+fn expand_event_parse_log(cx: &Context) -> TokenStream {
+    let all_events = {
+        let mut all_events = cx
+            .artifact
+            .abi
+            .events()
+            .map(|event| {
+                let struct_name = expand_struct_name(&event);
+
+                let name = Literal::string(&event.name);
+                let decode_event = quote! {
+                    log.decode(
+                        &Contract::artifact()
+                            .abi
+                            .event(#name)
+                            .expect("generated event decode")
+                    )
+                };
+
+                (event, struct_name, decode_event)
+            })
+            .collect::<Vec<_>>();
+
+        // NOTE: We sort the events by name so that the anonymous error decoding
+        //   is consistent. Since the events are stored in a `HashMap`, there is
+        //   no guaranteed order, and in the case where there is ambiguity in
+        //   decoding anonymous events, its nice if they follow some strict and
+        //   predictable order.
+        all_events.sort_unstable_by_key(|(event, _, _)| &event.name);
+        all_events
+    };
+
+    let standard_event_match_arms = all_events
+        .iter()
+        .filter(|(event, _, _)| !event.anonymous)
+        .map(|(event, struct_name, decode_event)| {
+            // These are all possible stardard (i.e. non-anonymous) events that
+            // the contract can produce, along with its signature and index in
+            // the contract ABI. For these, we match topic 0 to the signature
+            // and try to decode.
+
+            let signature = expand_hash(event.signature());
+            quote! {
+                #signature => Ok(Event::#struct_name(#decode_event?)),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let anonymous_event_try_decode = all_events
+        .iter()
+        .filter(|(event, _, _)| event.anonymous)
+        .map(|(_, struct_name, decode_event)| {
+            // For anonymous events, just try to decode one at a time and return
+            // the first that succeeds.
+
+            quote! {
+                if let Ok(data) = #decode_event {
+                    return Ok(Event::#struct_name(data));
+                }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let invalid_data = expand_invalid_data();
+
+    quote! {
+        impl self::ethcontract::contract::ParseLog for Event {
+            fn parse_log(
+                log: self::ethcontract::RawLog,
+            ) -> Result<Self, self::ethcontract::errors::ExecutionError> {
```

Comment:
I understand why this needs to return an error now, but I still don't get why this is not using `std::convert::TryFrom`? Does it have to do with the fact that this trait is a subtrait of Sized?

Reply:
`TryFrom` is also [`Sized`](https://doc.rust-lang.org/src/core/convert/mod.rs.html#488). There is no technical reason to not use `TryFrom`, but it would have to be a `TryFrom` that restricts the error type `TryFrom<RawLog, Error = ExecutionError>`.

The main motivation was that I feel like the `TryFrom` trait is a slightly different abstraction. From the docs:
> Simple and safe type conversions that may fail in a controlled way under some circumstances. It is the reciprocal of `TryInto`.
>
> This is useful when you are doing a type conversion that may trivially succeed but may also need special handling.

My understanding of this is that it is for trivial conversions with edge cases and not things like parsing. Its kind of like `FromStr`, which could in many cases be modeled as `TryFrom<&str>` (although the later has slightly different lifetime semantics).

Reply:
I don't feel strongly about this, so if you prefer using `TryFrom` here I don't really have any objections. Let me know @fleupold .

Reply:
I have no strong feelings either and since it's internal it probably doesn't make a difference.